api/pkg/apis/v1alpha1/vendors/campaigns-container-vendor.go

```diff
@@ -0,0 +1,163 @@
+			Parameters: []string{"name?"},
+		},
+	}
+}
+
+func (c *CampaignContainersVendor) onCampaignContainers(request v1alpha2.COARequest) v1alpha2.COAResponse {
+	pCtx, span := observability.StartSpan("onCampaignContainers", request.Context, &map[string]string{
+		"method": "onCampaignContainers",
+	})
+	defer span.End()
+	ccLog.Infof("V (CampaignContainers): onCampaignContainers, method: %s, traceId: %s", request.Method, span.SpanContext().TraceID().String())
+
+	id := request.Parameters["__name"]
+	namespace, exist := request.Parameters["namespace"]
+	if !exist {
+		namespace = constants.DefaultScope
+	}
+
+	ccLog.Infof("V (CampaignContainers): onCampaignContainers, method: %s, traceId: %s", string(request.Method), span.SpanContext().TraceID().String())
+	switch request.Method {
+	case fasthttp.MethodGet:
+		ctx, span := observability.StartSpan("onCampaignContainers-GET", pCtx, nil)
+		var err error
+		var state interface{}
+		isArray := false
+		if id == "" {
+			// Change partition back to empty to indicate ListSpec need to query all namespaces
+			if !exist {
+				namespace = ""
+			}
+			state, err = c.CampaignContainersManager.ListState(ctx, namespace)
+			isArray = true
+		} else {
+			state, err = c.CampaignContainersManager.GetState(ctx, id, namespace)
+		}
+		if err != nil {
+			ccLog.Errorf("V (CampaignContainers): onCampaignContainers failed - %s, traceId: %s", err.Error(), span.SpanContext().TraceID().String())
+			return observ_utils.CloseSpanWithCOAResponse(span, v1alpha2.COAResponse{
+				State: v1alpha2.InternalError,
+				Body:  []byte(err.Error()),
+			})
+		}
+		jData, _ := utils.FormatObject(state, isArray, request.Parameters["path"], request.Parameters["doc-type"])
+		resp := observ_utils.CloseSpanWithCOAResponse(span, v1alpha2.COAResponse{
+			State:       v1alpha2.OK,
+			Body:        jData,
+			ContentType: "application/json",
+		})
+		if request.Parameters["doc-type"] == "yaml" {
+			resp.ContentType = "application/text"
+		}
+		return resp
+	case fasthttp.MethodPost:
+		ctx, span := observability.StartSpan("onCampaignContainers-POST", pCtx, nil)
+		campaign := model.CampaignContainerState{
+			ObjectMeta: model.ObjectMeta{
+				Name:      id,
+				Namespace: namespace,
+			},
+			Spec: &model.CampaignContainerSpec{},
+		}
+
+		err := c.CampaignContainersManager.UpsertState(ctx, id, campaign)
+		if err != nil {
+			ccLog.Infof("V (CampaignContainers): onCampaignContainers failed - %s, traceId: %s", err.Error(), span.SpanContext().TraceID().String())
+			return observ_utils.CloseSpanWithCOAResponse(span, v1alpha2.COAResponse{
+				State: v1alpha2.InternalError,
+				Body:  []byte(err.Error()),
+			})
+		}
+
+		return observ_utils.CloseSpanWithCOAResponse(span, v1alpha2.COAResponse{
+			State: v1alpha2.OK,
+		})
+	case fasthttp.MethodDelete:
+		ctx, span := observability.StartSpan("onCampaignContainers-DELETE", pCtx, nil)
+		err := c.CampaignContainersManager.DeleteState(ctx, id, namespace)
```

Comment:
What happens to the campaigns under the container? Maybe to use a Campaign manager to check children before deletion is allowed?

Reply:
Maybe we can register container webhooks to guard deletion if children count is not empty. Then the behavior via AMR or kubectl in local system will be aligned. Then back to consider API users, if we would guard the same, need to update manager to block deletion but API users behavior will not block any cases right now. So we can leave container manager clean without children check? What do you think?
TODO will be:
- Add container webhooks
- Validate deletion: check children whose ownerReference has the current container uid.
- Leave container the same.

Reply:
Updated in the latest commit. Container webhook added to validate children resources for deletion.